File: Python/buffer.py
class CircularBuffer:
    def __init__(self, capacity):
        self.capacity = capacity
        self.data = [0] * capacity
        self.read_index = 0
        self.write_index = 0
        self.size = 0
    
    def reset(self):
        self.read_index = 0
        self.write_index = 0
        self.size = 0
        self.data = [0] * self.capacity
    
    def length(self):
        return self.size
    
    def write(self, value):
        if self.size >= self.capacity:
            self.read_index = (self.read_index + 1) % self.capacity
            self.size -= 1
        
        self.data[self.write_index] = value
        self.write_index = (self.write_index + 1) % self.capacity
        self.size += 1
    
    def pop_data(self):
        if self.size == 0:
            return None  # Buffer empty
        
        value = self.data[self.read_index]
        self.read_index = (self.read_index + 1) % self.capacity
        self.size -= 1
        return value


class ClipBuffer:
    def __init__(self, capacity):
        self.capacity = capacity
        self.data = [0] * capacity
        self.size = 0
    
    def reset(self):
        self.data = [0] * self.capacity
        self.size = 0
    

def pop_all_data(circular_buffer, clip_buffer):
    if circular_buffer.size == 0:
        clip_buffer.size = 0
        return 1  # Buffer empty
    
    count = min(circular_buffer.size, clip_buffer.capacity)
    
    for i in range(count):
        clip_buffer.data[i] = circular_buffer.pop_data()
    
    clip_buffer.size = count
    return 0

File: Java/Buffer.java
public class Buffer {

    // Circular buffer class
    public static class CircularBuffer {
        private byte[] data;
        private int capacity;
        private int size;
        private int readIndex;
        private int writeIndex;

        public CircularBuffer(int capacity) {
            this.capacity = capacity;
            this.data = new byte[capacity];
            this.size = 0;
            this.readIndex = 0;
            this.writeIndex = 0;
        }

        public void init(byte[] bufferArray) {
            this.data = bufferArray;
            this.capacity = bufferArray.length;
            reset();
        }

        public void reset() {
            this.readIndex = 0;
            this.writeIndex = 0;
            this.size = 0;
            java.util.Arrays.fill(data, (byte) 0);
        }

        public int length() {
            return size;
        }

        public void write(byte data) {
            if (size >= capacity) {
                readIndex = (readIndex + 1) % capacity;
                size--;
            }
            this.data[writeIndex] = data;
            writeIndex = (writeIndex + 1) % capacity;
            size++;
        }

        public boolean popAllData(ClipBuffer clip) {
            if (size == 0) {
                clip.setSize(0);
                return true;
            }

            int count = Math.min(size, clip.getCapacity());
            for (int i = 0; i < count; i++) {
                clip.getData()[i] = this.data[readIndex];
                readIndex = (readIndex + 1) % capacity;
            }

            size -= count;
            clip.setSize(count);
            return false;
        }
    }

    // Clip buffer class
    public static class ClipBuffer {
        private byte[] data;
        private int capacity;
        private int size;

        public ClipBuffer(int capacity) {
            this.capacity = capacity;
            this.data = new byte[capacity];
            this.size = 0;
        }

        public void init(byte[] clipArray) {
            this.data = clipArray;
            this.capacity = clipArray.length;
            reset();
        }

        public void reset() {
            java.util.Arrays.fill(data, (byte) 0);
            size = 0;
        }

        public int getSize() {
            return size;
        }

        public void setSize(int size) {
            this.size = size;
        }

        public int getCapacity() {
            return capacity;
        }

        public byte[] getData() {
            return data;
        }
    }
}
